Implement AI commit message core functionality plus improved error handling and config management

ilgaur opened the pull request:
This PR implements the core functionality for ai-commiter with enhanced reliability and maintainability. after closing a bloated PR I made before for the initial development, I have created this one that sums up all the development and minor changes and refactors as they have been discussed.

Core stuff that I've done:
- Pre-commit integration: Uses pre-commit framework for hook orchestration
- AI-powered commit messages: Integrates with OpenRouter API
- Staged diff processing: Captures and processes git staged changes via subprocess
- Semantic commit format: Enforces conventional commit format `type(scope): description`
- Flexible configuration: Environment-based config with sensible defaults

Technical Improvements and refactors that have been done after comments in #4, for some of the comments, I have created their respective issues and sub-issues, which are #3 and #8 :
- Better error handling: Proper exception raising
- Modular config management: Separated environment variable retrieval from config creation
- Enhanced reliability: Better error handling throughout the commit generation flow
- Better maintainability: Improved code structure and documentation

Closes some issues from #3 and #8

ilgaur committed:
chore: init hook script

ilgaur committed:
chore: create pre-commit-hook to define repo hooks

ilgaur committed:
chore: add setup.py

ilgaur committed:
chore: use commit-msg hook

ilgaur committed:
chore: add overall structure

ilgaur committed:
chore: remove setup.py

ilgaur committed:
chore: remove hook.py

ilgaur committed:
chore: skeleton implementation

ilgaur committed:
chore: add validation for argument count

ilgaur committed:
chore: add the command for getting the diff data in the get_diff()

ilgaur committed:
fix: get cleaner output in the get_diff

ilgaur committed:
chore: add seperate function for generating commit messages

ilgaur committed:
chore: add client initialization + api request for generating commit messages to the generate_commit_msg()

ilgaur committed:
fix: return the completion content for the generate commit message func and include it in the function orchestration in the main

ilgaur committed:
chore: add get_api_config() to get needed creds via subprocess + get_api_config() inclusion in the generate_commit_msg() + minor syntax fix in the commit_messae validation in the main()

ilgaur committed:
chore: add check for api config + create comprehensive prompts for commit message generation

ilgaur committed:
chore: handle commit message writing with error handling and check for empty diff data

ilgaur committed:
docs: added comment for testing the hook

ilgaur committed:
fix: cleaner block ordering. Fixes #5 + add openai dependency. Fixes: #6

ilgaur committed:
fix: use explicit returns instaed of FAIL and PASS. Fixes: #7

ilgaur committed:
fix: add additional dependencies to the .pre-commit-hooks.yaml

ilgaur committed:
chore: add explanations in the readme + add the .pre-commit-config.yaml to the gitignore

ilgaur committed:
chore: refactor config env var handling using a data class and populate it with the appropriate env vars using os.environ.get  + have some optional env vars and use defaults on them so users can override. Addresses: #9

ilgaur committed:
refactor: improve error handling and separation of concerns in hook.py

- Split get_api_config() into get_config_env_vars() and create_config_from_env_vars()
- Replace print statements with proper exception raising for better error propagation
- Add comprehensive try-catch blocks in main() function
- Pass config as parameter to generate_commit_msg() instead of creating it internally
- Improve error messages with more descriptive context
- Add comments for better code documentation

This refactoring makes the code more modular, testable, and provides better error handling throughout the commit message generation flow.

tavallaie commented on pyproject.toml:
```diff
@@ -0,0 +1,17 @@
+description = "Create commit messages using AI"
+readme = "README.md"
+requires-python = ">=3.12"
+dependencies = ["openai"]
```
we need openai version

tavallaie commented on ai_commiter/hook.py:
```diff
@@ -0,0 +1,117 @@
+        return 1
+
+if __name__ == '__main__':
+    sys.exit(main())
```
why not use main() directly?

tavallaie commented on ai_commiter/hook.py:
```diff
@@ -0,0 +1,117 @@
+        return result
+    except Exception as e:
+        print(f"Error: {e}")
+        return 1
```
try not returning unnecessary variable or 1

tavallaie commented on ai_commiter/hook.py:
```diff
@@ -0,0 +1,117 @@
+def main():
+    #first arg is the script name and the second arg is the file which I'm validating to make sure it exists
+    if len(sys.argv) < 2:
+        return 1
```
one idea makes it a function for more readability or use break and continue

tavallaie commented on ai_commiter/hook.py:
```diff
@@ -0,0 +1,117 @@
+            raise Exception("API Key is missing")
+        if not env_vars["base_url"]:
+            raise Exception("Base URL is missing")
+
```
please make `config` object and return that varible

tavallaie commented on ai_commiter/hook.py:
```diff
@@ -0,0 +1,117 @@
+            "user_prompt": os.environ.get("AI_COMMITER_USER_PROMPT", DEFAULT_USER_PROMPT)
+        }
+    except Exception as e:
+        raise Exception(f"Error retrieving environment variables: {e}")
```
please clean this function to not have extra `try`

tavallaie commented on ai_commiter/hook.py:
```diff
@@ -0,0 +1,117 @@
+def get_diff():
+    #returns diff data
+    result = subprocess.run(['git', 'diff', '--staged'], capture_output=True, text=True).stdout
+    return result
```
use `try` here

ilgaur committed:
refactor: extract configuration constants and improve error handling

  - Move DEFAULT_SYSTEM_PROMPT, DEFAULT_USER_PROMPT, and DEFAULT_MODEL to separate config module
  - Add proper exception handling with try-catch blocks in get_diff()
  - Change environment variable retrieval to return None instead of empty strings for missing values
  - Extract argument validation into separate validate_arguments() function
  - Simplify main() function by removing nested try-catch and using explicit error raising
  - Remove return value from write_commit_message() function
  - Update function comments and structure for better clarity
